app/javascript/components/widget-transition-controls.jsx

```diff
@@ -0,0 +1,166 @@
+import React, { useEffect, useReducer } from 'react';
+import PropTypes from 'prop-types';
+import { mathMod } from 'ramda';
+
+const prefetchDelay = 1500;
+
+export const selectShouldRequestWidget = state =>
+  state.msRemainingSinceLastStart < Date.now() - state.startTimestamp;
+
+export const selectShouldRequestPrefetch = state =>
+  state.noPrefetchRequestedYet &&
```

Comment:
Nit: "keep it positive" is generally a good mantra for boolean names. `needsPrefetch`.